src/source/CppMarshal.scala

```diff
@@ -113,8 +124,17 @@ class CppMarshal(spec: Spec) extends Marshal(spec) {
 
   def include(ident: String): String = q(spec.cppIncludePrefix + spec.cppFileIdentStyle(ident) + "." + spec.cppHeaderExt)
 
-  private def toCppType(ty: TypeRef, namespace: Option[String] = None): String = toCppType(ty.resolved, namespace)
-  private def toCppType(tm: MExpr, namespace: Option[String]): String = {
+  private def toCppType(ty: TypeRef, scopeNames: Seq[String] = Seq(), namespace: Option[String] = None): String =
+    toCppType(ty.resolved, scopeNames, namespace)
+
+  private def toCppType(tm: MExpr, scopeNames: Seq[String], namespace: Option[String]): String = {
+    def withNamespace(name: String): String = {
+      val ns = namespace match {
+        case Some(ns) => Some(ns)
+        case None => if (scopeNames.contains(name)) Some(spec.cppNamespace) else None
```

Comment:
There's an assumption here that there are never unqualified symbols which aren't for generated types.  I think that's a safe assumption currently, at least within the C++ generator (JNI and ObjC use plenty of global names), but it worries me a bit, and is worth at least documenting.